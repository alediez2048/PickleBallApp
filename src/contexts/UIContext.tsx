import React, { createContext, useContext, useReducer, useCallback, useEffect } from 'react';
import { ColorSchemeName, useColorScheme as useRNColorScheme } from 'react-native';

// State interface
interface UIState {
  colorScheme: ColorSchemeName;
  isLoading: boolean;
  toast: {
    visible: boolean;
    message: string;
    type: 'success' | 'error' | 'info';
  };
}

// Action types
type UIAction =
  | { type: 'SET_COLOR_SCHEME'; payload: ColorSchemeName }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SHOW_TOAST'; payload: { message: string; type: 'success' | 'error' | 'info' } }
  | { type: 'HIDE_TOAST' };

// Initial state
const initialState: UIState = {
  colorScheme: 'light',
  isLoading: false,
  toast: {
    visible: false,
    message: '',
    type: 'info',
  },
};

// Reducer
function uiReducer(state: UIState, action: UIAction): UIState {
  switch (action.type) {
    case 'SET_COLOR_SCHEME':
      return { ...state, colorScheme: action.payload };
    case 'SET_LOADING':
      return { ...state, isLoading: action.payload };
    case 'SHOW_TOAST':
      return {
        ...state,
        toast: {
          visible: true,
          message: action.payload.message,
          type: action.payload.type,
        },
      };
    case 'HIDE_TOAST':
      return {
        ...state,
        toast: {
          ...state.toast,
          visible: false,
        },
      };
    default:
      return state;
  }
}

// Context interface
interface UIContextType extends UIState {
  setColorScheme: (scheme: ColorSchemeName) => void;
  setLoading: (isLoading: boolean) => void;
  showToast: (message: string, type: 'success' | 'error' | 'info') => void;
  hideToast: () => void;
  useThemeColor: (props: { light?: string; dark?: string }, colorName: string) => string;
}

// Create context
const UIContext = createContext<UIContextType | undefined>(undefined);

// Provider component
export function UIProvider({ children }: { children: React.ReactNode }) {
  const [state, dispatch] = useReducer(uiReducer, initialState);
  const systemColorScheme = useRNColorScheme();

  // Sync with system color scheme
  useEffect(() => {
    dispatch({ type: 'SET_COLOR_SCHEME', payload: systemColorScheme });
  }, [systemColorScheme]);

  const setColorScheme = useCallback((scheme: ColorSchemeName) => {
    dispatch({ type: 'SET_COLOR_SCHEME', payload: scheme });
  }, []);

  const setLoading = useCallback((isLoading: boolean) => {
    dispatch({ type: 'SET_LOADING', payload: isLoading });
  }, []);

  const showToast = useCallback((message: string, type: 'success' | 'error' | 'info') => {
    dispatch({ type: 'SHOW_TOAST', payload: { message, type } });
    // Auto-hide toast after 3 seconds
    setTimeout(() => {
      dispatch({ type: 'HIDE_TOAST' });
    }, 3000);
  }, []);

  const hideToast = useCallback(() => {
    dispatch({ type: 'HIDE_TOAST' });
  }, []);

  const useThemeColor = useCallback(
    (props: { light?: string; dark?: string }, colorName: string): string => {
      const theme = state.colorScheme ?? 'light';
      const colorFromProps = props[theme];

      if (colorFromProps) {
        return colorFromProps;
      }

      // Default theme colors
      const Colors = {
        light: {
          text: '#000000',
          background: '#ffffff',
          primary: '#2f95dc',
          secondary: '#f4f4f5',
          tint: '#2f95dc',
          tabIconDefault: '#cccccc',
          tabIconSelected: '#2f95dc',
        },
        dark: {
          text: '#ffffff',
          background: '#000000',
          primary: '#2f95dc',
          secondary: '#27272a',
          tint: '#ffffff',
          tabIconDefault: '#cccccc',
          tabIconSelected: '#ffffff',
        },
      };

      return Colors[theme][colorName as keyof typeof Colors.light];
    },
    [state.colorScheme]
  );

  const value = {
    ...state,
    setColorScheme,
    setLoading,
    showToast,
    hideToast,
    useThemeColor,
  };

  return <UIContext.Provider value={value}>{children}</UIContext.Provider>;
}

// Hook for using the context
export function useUI() {
  const context = useContext(UIContext);
  if (context === undefined) {
    throw new Error('useUI must be used within a UIProvider');
  }
  return context;
} 