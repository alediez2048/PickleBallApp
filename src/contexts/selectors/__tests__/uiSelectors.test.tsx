import React from 'react';
import renderer from 'react-test-renderer';
import { UIProvider } from '../../UIContext';
import {
  useThemeMode,
  useLoadingState,
  useToastState,
  useThemedColor,
} from '../uiSelectors';
import { ColorSchemeName } from 'react-native';

interface MockUIContext {
  colorScheme: ColorSchemeName;
  isLoading: boolean;
  toast: {
    visible: boolean;
    message: string;
    type: 'success' | 'error' | 'info';
  };
  useThemeColor: (
    props: { light?: string; dark?: string },
    colorName: string
  ) => string;
}

const mockUIContext: MockUIContext = {
  colorScheme: 'light',
  isLoading: false,
  toast: {
    visible: false,
    message: '',
    type: 'info',
  },
  useThemeColor: (props, colorName): string => 
    props[mockUIContext.colorScheme as 'light' | 'dark'] || '#000000',
};

jest.mock('../../UIContext', () => ({
  UIProvider: ({ children }: { children: React.ReactNode }) => children,
  useUI: () => mockUIContext,
}));

// Create components to test each hook
const ThemeModeComponent: React.FC = () => {
  const themeMode = useThemeMode();
  return (
    <div data-testid="theme-mode">
      <div>isDark: {themeMode.isDark ? 'true' : 'false'}</div>
      <div>isLight: {themeMode.isLight ? 'true' : 'false'}</div>
      <div>current: {themeMode.current}</div>
    </div>
  );
};

const LoadingStateComponent: React.FC = () => {
  const loadingState = useLoadingState();
  return (
    <div data-testid="loading-state">
      <div>isLoading: {loadingState.isLoading ? 'true' : 'false'}</div>
      <div>isReady: {loadingState.isReady ? 'true' : 'false'}</div>
    </div>
  );
};

const ToastStateComponent: React.FC = () => {
  const toastState = useToastState();
  return (
    <div data-testid="toast-state">
      <div>visible: {toastState.visible ? 'true' : 'false'}</div>
      <div>message: {toastState.message}</div>
      <div>type: {toastState.type}</div>
      <div>isSuccess: {toastState.isSuccess ? 'true' : 'false'}</div>
      <div>isError: {toastState.isError ? 'true' : 'false'}</div>
      <div>isInfo: {toastState.isInfo ? 'true' : 'false'}</div>
    </div>
  );
};

interface ThemedColorProps {
  colorName: string;
  lightColor?: string;
  darkColor?: string;
}

const ThemedColorComponent: React.FC<ThemedColorProps> = ({ 
  colorName, 
  lightColor,
  darkColor
}) => {
  const color = useThemedColor(colorName, lightColor, darkColor);
  return (
    <div data-testid="themed-color" style={{ color }}>
      {color}
    </div>
  );
};

describe('UI Selectors', () => {
  const wrapper = (component: React.ReactElement) => (
    <UIProvider>{component}</UIProvider>
  );

  describe('useThemeMode', () => {
    it('returns correct theme mode state for light theme', () => {
      mockUIContext.colorScheme = 'light';
      const tree = renderer.create(wrapper(<ThemeModeComponent />)).toJSON();
      expect(tree).toMatchSnapshot('light theme');
    });

    it('returns correct theme mode state for dark theme', () => {
      mockUIContext.colorScheme = 'dark';
      const tree = renderer.create(wrapper(<ThemeModeComponent />)).toJSON();
      expect(tree).toMatchSnapshot('dark theme');
    });
  });

  describe('useLoadingState', () => {
    it('returns correct loading state when not loading', () => {
      mockUIContext.isLoading = false;
      const tree = renderer.create(wrapper(<LoadingStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('not loading');
    });

    it('returns correct loading state when loading', () => {
      mockUIContext.isLoading = true;
      const tree = renderer.create(wrapper(<LoadingStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('loading');
    });
  });

  describe('useToastState', () => {
    it('returns correct toast state when toast is hidden', () => {
      mockUIContext.toast = {
        visible: false,
        message: '',
        type: 'info',
      };
      const tree = renderer.create(wrapper(<ToastStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('hidden toast');
    });

    it('returns correct toast state for success toast', () => {
      mockUIContext.toast = {
        visible: true,
        message: 'Success message',
        type: 'success',
      };
      const tree = renderer.create(wrapper(<ToastStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('success toast');
    });
  });

  describe('useThemedColor', () => {
    it('returns correct color for light theme', () => {
      mockUIContext.colorScheme = 'light';
      const tree = renderer.create(
        wrapper(<ThemedColorComponent colorName="background" lightColor="#ffffff" darkColor="#000000" />)
      ).toJSON();
      expect(tree).toMatchSnapshot('light mode color');
    });

    it('returns correct color for dark theme', () => {
      mockUIContext.colorScheme = 'dark';
      const tree = renderer.create(
        wrapper(<ThemedColorComponent colorName="background" lightColor="#ffffff" darkColor="#000000" />)
      ).toJSON();
      expect(tree).toMatchSnapshot('dark mode color');
    });

    it('returns default color when no custom colors provided', () => {
      const tree = renderer.create(
        wrapper(<ThemedColorComponent colorName="background" />)
      ).toJSON();
      expect(tree).toMatchSnapshot('default themed color');
    });
  });
}); 