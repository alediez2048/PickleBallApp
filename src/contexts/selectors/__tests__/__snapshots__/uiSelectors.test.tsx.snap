// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Selectors useLoadingState returns correct loading state when loading: loading 1`] = `null`;

exports[`UI Selectors useLoadingState returns correct loading state when not loading: not loading 1`] = `null`;

exports[`UI Selectors useThemeMode returns correct theme mode state for dark theme: dark theme 1`] = `null`;

exports[`UI Selectors useThemeMode returns correct theme mode state for light theme: light theme 1`] = `null`;

exports[`UI Selectors useThemedColor returns correct color for dark theme: dark mode color 1`] = `null`;

exports[`UI Selectors useThemedColor returns correct color for light theme: light mode color 1`] = `null`;

exports[`UI Selectors useThemedColor returns default color when no custom colors provided: default themed color 1`] = `null`;

exports[`UI Selectors useToastState returns correct toast state for success toast: success toast 1`] = `null`;

exports[`UI Selectors useToastState returns correct toast state when toast is hidden: hidden toast 1`] = `null`;
